apiserver/tools.go

```diff
@@ -20,20 +20,175 @@ import (
+// established that initiated this call helping identifying the who behind the
+// operation.
+func (a *agentBinaryStoreLogShim) AddAgentBinaryWithSHA256(
+	ctx context.Context,
+	data io.Reader,
+	version coreagentbinary.Version,
+	dataSize int64,
+	dataSHA256Sum string,
+) error {
+	logger.Infof(
+		ctx,
+		"agent binaries being added to %q for %q with sha %q on behalf of entity %q",
+		a.StoreName, version.String(), dataSHA256Sum, httpcontext.EntityForContext(ctx),
+	)
+	return a.AddAgentBinaryWithSHA256(ctx, data, version, dataSize, dataSHA256Sum)
+}
+
+// BlockCheckerGetterForServices returns a [BlockCheckerGetter] that is
+// constructed from the supplied context.
+func BlockCheckerGetterForServices(servicesGetter DomainServicesGetter) func(context.Context) (BlockChecker, error) {
+	return func(ctx context.Context) (BlockChecker, error) {
+		svc, err := servicesGetter(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		return common.NewBlockChecker(svc.BlockCommand()), nil
+	}
+}
+
+// controllerAgentBinaryStoreForHTTPContext provides a deferred getter that
+// will provide the controller's [AgentBinaryStore] for the given [httpContext].
+func controllerAgentBinaryStoreForHTTPContext(httpCtx httpContext) AgentBinaryStoreGetter {
+	return func(r *http.Request) (AgentBinaryStore, error) {
+		services, err := httpCtx.domainServicesForRequest(r.Context())
+		if err != nil {
+			return nil, internalerrors.Capture(err)
+		}
+
+		return &agentBinaryStoreLogShim{
+			AgentBinaryStore: services.ControllerAgentBinaryStore(),
+			StoreName:        "controller agent binary store",
+		}, nil
+	}
+}
+
+// migratingAgentBinaryStoreForHTTPContext provides a deferred getter that will
+// provide the agent binary store for the model that is being migrated as part
+// of the request.
+func migratingAgentBinaryStoreForHTTPContext(httpCtx httpContext) AgentBinaryStoreGetter {
+	return func(r *http.Request) (AgentBinaryStore, error) {
+		services, err := httpCtx.domainServicesDuringMigrationForRequest(r)
+		if err != nil {
+			return nil, internalerrors.Capture(err)
+		}
+
+		modelUUID, exists := httpcontext.MigrationRequestModelUUID(r)
+		if !exists {
+			modelUUID = "unknown"
+		}
+
+		return &agentBinaryStoreLogShim{
+			AgentBinaryStore: services.AgentBinaryStore(),
+			StoreName:        "model " + modelUUID,
+		}, nil
+	}
+}
+
+// modelAgentBinaryStoreForHTTPContext provides a deferred getter that will
+// provide the models [AgentBinaryStore] for the given [httpContext].
+func modelAgentBinaryStoreForHTTPContext(httpCtx httpContext) AgentBinaryStoreGetter {
+	return func(r *http.Request) (AgentBinaryStore, error) {
+		services, err := httpCtx.domainServicesForRequest(r.Context())
+		if err != nil {
+			return nil, internalerrors.Capture(err)
+		}
+
+		modelUUID, exists := httpcontext.RequestModelUUID(r.Context())
+		// This should never happen but it needs to be accounted for.
+		if !exists {
+			modelUUID = "unknown"
+		}
```

Comment:
IMO we return an error here. Not having a model uuid should be considered fatal and using an error will highlight any potential issue pretty quickly.

Reply:
Yeah this seems fair.